src/net/server/message.rs

```diff
@@ -0,0 +1,372 @@
+    /// message types require this, e.g. UDP DNS message length is determined
+    /// by the size of the UDP message received, while for TCP DNS messages
+    /// the number of bytes to expect is determined by the first two bytes
+    /// received.
+    const MIN_HDR_BYTES: usize;
+
+    /// The concrete type of message that we produce from given message
+    /// bytes.
+    type Msg;
+
+    /// The actual number of message bytes to follow given at least
+    /// MIN_HDR_BYTES of message header.
+    fn determine_msg_len(hdr_buf: &mut RequestOctets) -> usize;
+
+    /// Convert a sequence of bytes to a concrete message.
+    fn from_octets(octets: RequestOctets) -> Result<Self::Msg, ShortMessage>;
+}
+
+/// An implementation of MsgProvider for DNS [`Message`]s.
+impl<RequestOctets: AsRef<[u8]>> MsgProvider<RequestOctets>
+    for Message<RequestOctets>
+{
+    /// RFC 1035 section 4.2.2 "TCP Usage" says:
+    ///     "The message is prefixed with a two byte length field which gives
+    ///      the message length, excluding the two byte length field.  This
+    ///      length field allows the low-level processing to assemble a
+    ///      complete message before beginning to parse it."
+    const MIN_HDR_BYTES: usize = 2;
+
+    type Msg = Self;
+
+    #[must_use]
+    fn determine_msg_len(hdr_buf: &mut RequestOctets) -> usize {
+        u16::from_be_bytes(hdr_buf.as_ref().try_into().unwrap()) as usize
+    }
+
+    fn from_octets(octets: RequestOctets) -> Result<Self, ShortMessage> {
+        Self::from_octets(octets)
+    }
+}
+
+//------------ ContextAwareMessage -------------------------------------------
+
+/// A DNS message with additional properties describing its context.
+///
+/// DNS messages don't exist in isolation, they are received from somewhere or
+/// created by something. This type wraps a message with additional context
+/// about its origins so that decisions can be taken based not just on the
+/// message itself but also on the circumstances surrounding its creation and
+/// delivery.
+#[derive(Debug)]
+pub struct ContextAwareMessage<T> {
+    message: T,
+    received_over_tcp: bool,
+    client_addr: std::net::SocketAddr,
+}
+
+impl<T> ContextAwareMessage<T> {
+    pub fn new(
+        message: T,
+        received_over_tcp: bool,
+        client_addr: std::net::SocketAddr,
+    ) -> Self {
+        Self {
+            message,
+            received_over_tcp,
+            client_addr,
+        }
+    }
+
+    /// Was this message received via a TCP transport?
+    pub fn received_over_tcp(&self) -> bool {
+        self.received_over_tcp
+    }
+
+    /// From which IP address and port number was this message received?
+    pub fn client_addr(&self) -> std::net::SocketAddr {
+        self.client_addr
+    }
+
+    /// Exchange this wrapper for the inner message that it wraps.
+    pub fn into_inner(self) -> T {
+        self.message
+    }
+}
+
+impl<T> core::ops::Deref for ContextAwareMessage<T> {
```

Comment:
I am (mis)using Deref here because I want the caller to have access to all of the functionality of the underlying message without `ContextAwareMessage` or the caller having to care what that functionality is.